### PR TITLE
feat(`cheatcodes`): mockCall with bytes4 data

### LINKED ISSUE #9213: feat(`cheatcodes`): Consider accepting bytes4 as data input to vm.mockCall
`vm.mockCall` currently has the following function signatures:

```solidity
function mockCall(address where, bytes calldata data, bytes calldata retdata) external;
```

```solidity
function mockCall(
    address where,
    uint256 value,
    bytes calldata data,
    bytes calldata retdata
) external;
```

In its current state, this leads to an awkward syntax for using `vm.mockCall` for any call to a specific function signature. Foundry docs give the following example for how to do this:

```solidity
function testMockCall() public {
    vm.mockCall(
        address(0),
        abi.encodeWithSelector(MyToken.balanceOf.selector),
        abi.encode(10)
    );
    assertEq(IERC20(address(0)).balanceOf(address(1)), 10);
    assertEq(IERC20(address(0)).balanceOf(address(2)), 10);
}
```

This certainly works, but `abi.encodeWithSelector` has the semantic meaning of encoding a function call and is awkward when it's being deliberately abused as a way to basically just mean `abi.encodePacked(MyToken.balanceOf.selector)`. We could just change the docs to use `abi.encodePacked` but I think it would be beneficial to consider accepting two additional function signatures:

```solidity
function mockCall(address where, bytes4 data, bytes calldata retdata) external;
```

```solidity
function mockCall(
    address where,
    uint256 value,
    bytes4 data,
    bytes calldata retdata
) external;
```

This would mean we could improve the syntax to:

```solidity
function testMockCall() public {
    vm.mockCall(
        address(0),
        MyToken.balanceOf.selector,
        abi.encode(10)
    );
    assertEq(IERC20(address(0)).balanceOf(address(1)), 10);
    assertEq(IERC20(address(0)).balanceOf(address(2)), 10);
}
```

Which makes it much clearer what's going on.

### PR DESCRIPTION
**Pull request by yash-atreya**
## Motivation

Closes #9213 

## Solution

- Add overloads for mock cheatcodes with `bytes4 data` arg to pass the selector like `token.balanceOf.selector` instead of `abi.encodeWithSelector`

### CODE REVIEW

**Review by grandizzy — COMMENTED** [review]
I checked again the ticket and I think we got this wrong when discussed action, looks like the proposal is just to add 2 more sigs to make UX better but keep existings @zerosnacks could you also double check please
```Solidity
function mockCall(address where, bytes4 data, bytes calldata retdata) external;

function mockCall(
    address where,
    uint256 value,
    bytes4 data,
    bytes calldata retdata
) external;
```

**Comment by grandizzy** [comment]
looks good, I'll push a change to make the same for `vm.mockCallRevert` sigs and then merge if OK with you @yash-atreya

File: crates/cheatcodes/spec/src/vm.rs [hunk 1]
```diff
@@ -527,6 +527,22 @@ interface Vm {
...
+    /// Overload to pass the function selector directly `token.approve.selector` instead of `abi.encodeWithSelector(token.approve.selector)`.
+    #[cheatcode(group = Evm, safety = Unsafe)]
+    function mockCall(address callee, bytes4 data, bytes calldata returnData) external;
+
+    /// Mocks a call to an address with a specific `msg.value`, returning specified data.
+    /// Calldata match takes precedence over `msg.value` in case of ambiguity.
+    ///
+    /// Overload to pass the function selector directly `token.approve.selector` instead of `abi.encodeWithSelector(token.approve.selector)`.
+    #[cheatcode(group = Evm, safety = Unsafe)]
+    function mockCall(address callee, uint256 msgValue, bytes4 data, bytes calldata returnData) external;
```
Comment (zerosnacks) [inline crates/cheatcodes/spec/src/vm.rs:544]: Can we rename the `bytes4 data` to `bytes4 selector` to make it more explicit what its purpose is?

File: crates/cheatcodes/spec/src/vm.rs [hunk 2]
```diff
@@ -544,6 +560,19 @@ interface Vm {
...
+    /// Overload to pass the function selector directly `token.approve.selector` instead of `abi.encodeWithSelector(token.approve.selector)`.
+    #[cheatcode(group = Evm, safety = Unsafe)]
+    function mockCallRevert(address callee, bytes4 data, bytes calldata revertData) external;
+
+    /// Reverts a call to an address with a specific `msg.value`, with specified revert data.
+    ///
+    /// Overload to pass the function selector directly `token.approve.selector` instead of `abi.encodeWithSelector(token.approve.selector)`.
+    #[cheatcode(group = Evm, safety = Unsafe)]
+    function mockCallRevert(address callee, uint256 msgValue, bytes4 data, bytes calldata revertData)
+        external;
```
Comment (zerosnacks) [inline crates/cheatcodes/spec/src/vm.rs:574]: Same here

**Review by zerosnacks — APPROVED** [review]
Implementation LGTM, small note regarding name of `bytes4 data`

**Comment by smartcontracts** [comment]
Woot, thank you all!

**Pull request merged**